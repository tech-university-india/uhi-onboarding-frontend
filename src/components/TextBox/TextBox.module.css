.input-box { 
  position: relative; 
}

.partial-input { 
  padding-right: 6.25rem; 
  text-align: right;
}

.fixed-text { 
  position: absolute; 
  display: block; 
  right: 0.75rem; 
  top: 0.75rem; 
  z-index: 9; 
}