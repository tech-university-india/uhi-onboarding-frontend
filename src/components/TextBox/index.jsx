import React from 'react'
import PropTypes from 'prop-types'

function TextBox ({ onChange, placeholder, disabled, noBg, partialText, text }) {
  // The following are used to calculate the padding required by the text being entered
  // into the textbox in the partialText mode so that the entered text does not overlap
  // onto the partialText.

  const PAD_PARTIAL_TEXT_DOT = 0.625
  const PAD_PARTIAL_TEXT_NO_DOT = 0.68
  const PAD_PARTIAL_TEXT = `${partialText === undefined
    ? 0
    : (partialText?.length * (partialText?.startsWith('.') ? PAD_PARTIAL_TEXT_DOT : PAD_PARTIAL_TEXT_NO_DOT))}rem`

  return (
    <>
      {partialText !== undefined
        ? <div className="relative">
          <input type="text" disabled={disabled} className={`disabled:text-preFillText p-3 w-full rounded-lg ${disabled ? 'shadow-textBoxInset' : 'shadow-textBox'} ${noBg ? '' : 'bg-textBox'} text-right`}
            style={{
              paddingRight: PAD_PARTIAL_TEXT
            }}
            onChange={(event) => {
              const eventCopy = { ...event, target: { ...event.target, value: event.target.value + partialText } }
              onChange(eventCopy)
            }}
            placeholder={placeholder}
            value={text}
          />
          <span className={'absolute block right-3 top-3 z-10 text-preFillText'}>{partialText}</span>
        </div>
        : <input type="text" disabled={disabled} className={`disabled:text-preFillText p-3 w-full rounded-lg ${disabled ? 'shadow-textBoxInset' : 'shadow-textBox'} ${noBg ? '' : 'bg-textBox'}`}
          onChange={onChange}
          placeholder={placeholder}
          value={text}
        />}
    </>
  )
}

TextBox.propTypes = {
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  noBg: PropTypes.bool,
  partialText: PropTypes.string,
  text: PropTypes.string
}

TextBox.defaultProps = {
  noBg: false,
  placeholder: ''
}

export default TextBox
