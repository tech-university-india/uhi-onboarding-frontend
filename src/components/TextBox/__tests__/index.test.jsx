/* eslint-disable no-undef */
import React from 'react'
import TextBox from '..'
import { fireEvent, render, screen } from '@testing-library/react'

describe('TextBox', () => {
  it('should render the component using default props with placeholder \'Aadhaar Number\'', () => {
    const dummyPropsDefault = {
      placeholder: 'Aadhaar Number',
      onChange: () => {}
    }

    const renderedComponent = render(<TextBox {...dummyPropsDefault} />)
    expect(
      renderedComponent.getByPlaceholderText('Aadhaar Number')
    ).toBeTruthy()
  })

  it('should render the component using props with text \'Aadhaar Number\'', () => {
    const dummyPropsText = {
      text: 'Aadhaar Number',
      onChange: () => {}
    }

    const renderedComponent = render(<TextBox {...dummyPropsText} />)
    expect(renderedComponent.getByDisplayValue('Aadhaar Number')).toBeTruthy()
  })

  it('should render the component using props with disabled as true and text \'Aadhaar Number\'', () => {
    const dummyPropsDisabled = {
      disabled: true,
      text: 'Aadhaar Number',
      onChange: () => {}
    }

    const renderedComponent = render(
      <TextBox {...dummyPropsDisabled} />
    ).asFragment()
    expect(renderedComponent).toMatchSnapshot()
  })

  it('should render the component using props for fixed text and placeholder as \'Health ID\'', () => {
    const dummyPropsFixedText = {
      partialText: '.ayush@sbx',
      placeholder: 'Health ID',
      onChange: () => {}
    }

    const renderedComponent = render(
      <TextBox {...dummyPropsFixedText} />
    ).asFragment()
    expect(renderedComponent).toMatchSnapshot()
  })

  it('should render the component using props for no bg colour and placeholder as \'Aadhaar Number\'', () => {
    const dummyPropsNoBg = {
      noBg: true,
      onChange: () => {}
    }

    const renderedComponent = render(
      <TextBox {...dummyPropsNoBg} />
    ).asFragment()
    expect(renderedComponent).toMatchSnapshot()
  })

  it('should have input component as disabled when disabled true is passed in props', () => {
    const dummyPropsDisabledOnly = {
      disabled: true,
      placeholder: 'Aadhaar Number',
      onChange: () => {}
    }

    render(<TextBox {...dummyPropsDisabledOnly} />)
    const inputComponent = screen.getByPlaceholderText('Aadhaar Number')
    expect(inputComponent).toBeDisabled()
  })

  it('should call onChange callback passed as part of props when text changes', () => {
    const dummyPropsWithCallback = {
      placeholder: 'Aadhaar Number',
      onChange: jest.fn(),
      partialText: 'Hello'
    }
    // expect(dummyPropsWithCallback.onChange).not.toBeCalled()
    render(<TextBox {...dummyPropsWithCallback} />)
    const inputComponent = screen.getByTestId('partialText')
    fireEvent.change(inputComponent, { target: { value: '123456781234' } })
    expect(dummyPropsWithCallback.onChange).toBeCalled()
  })
  it('should render the component using props for no bg colour and placeholder as \'Aadhaar Number\'', () => {
    const dummyPropsNoBg = {
      noBg: true,
      onChange: () => {},
      placeholder: 'Aadhaar Number'
    }

    render(
      <TextBox {...dummyPropsNoBg} />
    )
    expect(
      screen.getByPlaceholderText('Aadhaar Number').className).not.toContain('bg-textBox')
  })
  it('should render the component using props for disabled and placeholder as \'Aadhaar Number\'', () => {
    const dummyPropsDisabled = {
      disabled: true,
      onChange: () => {},
      placeholder: 'Aadhaar Number'
    }

    render(
      <TextBox {...dummyPropsDisabled} />
    )
    expect(
      screen.getByPlaceholderText('Aadhaar Number').className).toContain('shadow-textBoxInset')
  })
})
