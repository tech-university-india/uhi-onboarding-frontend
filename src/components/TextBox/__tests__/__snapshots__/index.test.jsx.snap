// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox should render the component using props for fixed text and placeholder as 'Health ID' 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <input
      class="disabled:text-preFillText p-3 w-full rounded-lg shadow-textBox bg-textBox text-right"
      data-testid="partialText"
      placeholder="Health ID"
      style="padding-right: 6.25rem;"
      type="text"
      value=""
    />
    <span
      class="absolute block right-3 top-3 z-10 text-preFillText"
    >
      .ayush@sbx
    </span>
  </div>
</DocumentFragment>
`;

exports[`TextBox should render the component using props for no bg colour and placeholder as 'Aadhaar Number' 1`] = `
<DocumentFragment>
  <input
    class="disabled:text-preFillText p-3 w-full rounded-lg shadow-textBox "
    placeholder=""
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`TextBox should render the component using props with disabled as true and text 'Aadhaar Number' 1`] = `
<DocumentFragment>
  <input
    class="disabled:text-preFillText p-3 w-full rounded-lg shadow-textBoxInset bg-textBox"
    disabled=""
    placeholder=""
    type="text"
    value="Aadhaar Number"
  />
</DocumentFragment>
`;
